packages/myst-cli/src/session/session.ts

```diff
@@ -157,4 +164,36 @@ export class Session implements ISession {
     });
     return warnings;
   }
+
+  async jupyterSessionManager(): Promise<SessionManager | undefined> {
+    if (this._jupyterSessionManager !== null) {
+      return Promise.resolve(this._jupyterSessionManager);
+    }
+    try {
+      const partialServerSettings = await new Promise<JupyterServerSettings>(
+        async (resolve, reject) => {
+          if (process.env.JUPYTER_BASE_URL === undefined) {
+            resolve(
+              findExistingJupyterServer() ||
+                (await launchJupyterServer(this.contentPath(), this.log)),
+            );
+          } else {
+            resolve({
+              baseUrl: process.env.JUPYTER_BASE_URL,
+              token: process.env.JUPYTER_TOKEN,
+            });
+          }
+        },
+      );
+      const serverSettings = ServerConnection.makeSettings(partialServerSettings);
+      const kernelManager = new KernelManager({ serverSettings });
+      const manager = new SessionManager({ kernelManager, serverSettings });
+      // TODO: this is a race condition, even though we shouldn't hit if if this promise is actually awaited
```

Comment:
`await manager.ready` in turn does await the `kernelManager.ready` if that's what you mean

Reply:
callers of this function would need to know to `await (await jupyterSessionManager()).ready` unless you await `manager.ready` in here?

Reply:
In this instance it's that there's no logic to avoid creating two managers if one doesn't properly await the promise to `session.jupyterSessionManager()`. i.e., between the first await and finally setting `this._jupyterSessionManager`.